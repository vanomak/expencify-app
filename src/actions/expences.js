import uuid from "uuid";

export const addExpence = ({
                        description = '',
                        note = '',
                        amount = 0,
                        createdAt = 0
                    } = {}) => ({
    type: 'ADD_EXPENCE',
    expence: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

export const removeExpence = ( id ) => ({
    type: 'REMOVE_EXPENCE',
    id
});

export const editExpence = (id, updates ) => ({
    type: 'EDIT_EXPENCE',
    id,
    updates
});