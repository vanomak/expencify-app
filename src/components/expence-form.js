import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import  { SingleDatePicker } from 'react-dates';

class ExpenceForm extends React.Component {
    constructor (props) {
        super(props);
        this.state = {
            description: props.expence ? props.expence.description :'',
            note: props.expence ? props.expence.note :'',
            amount: props.expence ? (props.expence.amount / 100).toString() : '',
            createdAt: props.expence ? moment(props.expence.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ( {amount} ));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ( { createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ( { calendarFocused: focused }));
    };


    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ( { error: 'Please, provide description and amount' }));
        } else {
            this.setState(() => ( { error: undefined }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" placeholder='Descripion' autoFocus
                           value={this.state.description}
                           onChange={this.onDescriptionChange}
                    />
                    <input
                        type="number"
                        placeholder='Amount'
                        onChange={this.onAmountChange}
                        value={this.state.amount}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false }
                        id="datePicker"
                    />
                    <div>
                        <textarea placeholder='Add note here' value={this.state.note} onChange={this.onNoteChange} />
                    </div>
                    <button>Add expense</button>
                </form>
            </div>
        )
    }
}

export default ExpenceForm;