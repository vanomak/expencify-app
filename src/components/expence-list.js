import React from 'react';
import {connect} from 'react-redux';
import ExpenceListItem from './expence-list-item';
import selectExpences from '../selectors/expences';


export const ExpenceList = (props) => (
    <div>
        {
            props.expences.length === 0 ? (
                <p>No expences</p>
            ) : (
                props.expences.map((item) => {
                    return (
                        <ExpenceListItem key={item.id} {...item} />
                    )
                })
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expences: selectExpences(state.expences, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenceList);

