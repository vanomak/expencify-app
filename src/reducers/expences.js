// Expences reducer
const expencesReducerDefaultState = [];

export default (state = expencesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENCE':
            return [
                ...state,
                action.expence
            ];
        case 'SET_EXPENCES':
            return [
                ...action.expences
            ];
        case 'EDIT_EXPENCE':
            return state.map((expence) => {
                if (expence.id === action.id) {
                    return {
                        ...expence,
                        ...action.updates
                    }
                }
                else
                    return expence;
            });
        case 'REMOVE_EXPENCE':
            return state.filter(({ id }) => id !== action.id);
        default:
            return state;
    }
};

