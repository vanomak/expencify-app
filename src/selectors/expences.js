import moment from 'moment';

// Get visible expences
export default (expences = [], { text = '', sortBy = 'date', startDate = undefined, endDate = undefined}) => {
    return expences.filter((el) => {
        const createdAtMoment = moment(el.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = el.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};
