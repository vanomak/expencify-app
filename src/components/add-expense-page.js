import React from 'react';
import ExpenceForm from './expence-form';
import {connect} from 'react-redux';
import {addExpence} from "../actions/expences";

export class AddExpensePage extends React.Component {
    onSubmit = (expence) => {
        this.props.addExpence(expence);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Expence Form</h1>
                <ExpenceForm
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpence: (expence) => dispatch(addExpence(expence))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);