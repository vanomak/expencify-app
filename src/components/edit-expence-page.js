import React from 'react';
import ExpenceForm from './expence-form';
import {connect} from 'react-redux';
import {editExpence, removeExpence} from "../actions/expences";

export class EditExpencePage extends React.Component {
    onSubmit = (expence) => {
        this.props.editExpence(this.props.expence.id, expence);
        this.props.history.push('/');
    };

    onRemove = (expence) => {
        this.props.removeExpence(this.props.expence.id);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenceForm
                    expence={this.props.expence}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expence: state.expences.find((el) => el.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch) => ({
    editExpence: (id, expence) => dispatch(editExpence(id, expence)),
    removeExpence: (id) => dispatch(removeExpence(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpencePage);