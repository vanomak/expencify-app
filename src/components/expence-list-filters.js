import React from 'react';
import {connect} from 'react-redux';
import selectExpences from "../selectors/expences";
import {setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate} from "../actions/filters";
import 'react-dates/initialize';
import  { DateRangePicker } from 'react-dates';
import {editExpence, removeExpence} from "../actions/expences";
// import  'react-dates/lib/css/_datepicker.css';

export class ExpenceListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };

    onDatesChange = ( {startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        if (e.target.value === 'date')
            this.props.sortByDate();
        else
            this.props.sortByAmount();
    };

    render() {
        const props = this.props;
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input className="text-input" placeholder="Search expences" type="text" value={props.filters.text} onChange={this.onTextChange} />
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={props.filters.sortBy} onChange={this.onSortChange}>
                            <option value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                            startDateId="start_date_id" // PropTypes.string.isRequired,
                            endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                            endDateId="end_date_id" // PropTypes.string.isRequired,
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                            showClearDates={true}
                            displayFormat="DD/MM/YYYY"
                            isOutsideRange={(day) => false }
                            onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                        />
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date))
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenceListFilters);
