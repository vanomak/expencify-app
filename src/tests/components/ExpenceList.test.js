import React from 'react';
import { shallow } from 'enzyme';
import { ExpenceList } from "../../components/expence-list";
import expences from '../fixtures/expences';

test('should render ExpenceList correctly with expences', () => {
    const wrapper = shallow(<ExpenceList expences={expences}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenceList correctly with empty message', () => {
    const wrapper = shallow(<ExpenceList expences={[]}/>);
    expect(wrapper).toMatchSnapshot();
});

