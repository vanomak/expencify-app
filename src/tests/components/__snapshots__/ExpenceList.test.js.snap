// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenceList correctly with empty message 1`] = `
<div>
  <p>
    No expences
  </p>
</div>
`;

exports[`should render ExpenceList correctly with expences 1`] = `
<div>
  <ExpenceListItem
    amount={770}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenceListItem
    amount={100}
    createdAt={-172800000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenceListItem
    amount={10000}
    createdAt={172800000}
    description="Credit card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
